lnrpc/invoicerpc/invoice.proto

```diff
@@ -244,28 +352,34 @@ message ListInvoiceResponse {
 message InvoiceSubscription {
     /**
     If specified (non-zero), then we'll first start by sending out
-    notifications for all added indexes with an add_index greater than this
-    value. This allows callers to catch up on any events they missed while they
+    all invoices that have an last_update_index greater than this value. This allows
```

Comment:
>doesn't require a scan of the entire db for each new client (would need to decode+check O(n) invoices, even if client only missed a few updates)

this isn't necessary, because of the `update_index` bucket, which is sorted? it looks up the `update_index` specified by the `subscribeinvoices` client and then uses a cursor to go through all the changed invoices until the end of the bucket is reached.

Reply:
Ahh okay, I was thinking the invoices themselves would be mutated, and we'd just search them all for ones that contain `last_update_index >= last_client_update_index`. Is your thinking we'd store a complete invoice for each `update_index`?

Reply:
From your comments above, would we garbage collect a prior `update_index` for an invoice that gets updated?

Reply:
Okay, I think i've got a picture of what you're describing. The `update_index` only stores the `invoice_number`, the invoice itself gets mutated directly and we can use the `update_index` to scan for any invoices that have been recently modified.

Indeed I like this and agree it's perfectly feasible. From an API perspsective it seems the only real tradeoff between the two is whether we expose intermediate state transitions to the client.

I wonder how much more complex the client state machine would have to be to handle these gaps. Say the invoice goes through states A -> B -> C -> D. If all intermediate transitions are preserved, the client only needs to handle A -> B, B -> C, and C -> D since they'll be replayed reliably.

If we skip intermediate states, the client also has to be ready to handle A -> C, A -> D, and B -> D, which definitely increases number of transitions that need to implemented and tested.

There is also information lost via aggregation, e.g. the client can't distinguish between the OPEN -> CANCEL and OPEN -> ACCEPTED -> CANCELED if they only see the first and last states.

Reply:
Yes exactly, it stores just a pointer. It is indeed a trade-off to keep the db simple. Actually, the concept stays identical to what it is currently. We already have an index bucket like that for add and settle. Just add one more for update and that's it.

I don't think it is a problem if we don't see the intermediate states. It is also a way to offer the information on the right abstraction level. I can imagine with multi-path payments, that we are really not willing to send updates on all the intermediate states.

Reply:
> I don't think it is a problem if we don't see the intermediate states.

If we're moving towards a design where invoice settling and switch decisions will be made off of these same subscriptions, I think it will be more reliable to replay all intermediate transitions. 

My preference would be to keep the model here closer to that of an RSM, since it makes everything easier to reason about and well defined. Otherwise the data that can be consumed during a restart is different from what can be presented in the transient case, which I think will be more complex in the long run.

Is it worth having non-deterministic behavior during restarts for mission-critical functionality?

> It is also a way to offer the information on the right abstraction level. I can imagine with multi-path payments, that we are really not willing to send updates on all the intermediate states.

I don't think this has to do with abstraction level. In the multi-path example, we would choose only to expose a state transition  (SETTLE/ACCEPT) once all of the HTLCs have arrived. I think this is only concerned with: should clients see all intermediary transitions, given that we've already decided they were relevant enough to expose in the first place.

Reply:
Seems the only difference is whether we store:
```
update_index => invoice number
```
vs
```
update_index => InvoiceUpdate {
        invoice_number
        update_index
        state
        settle_date
        amt_paid
}
```

Reply:
If we choose to have different semantics via rpc vs internal subscriptions, I don't think this poses too much overhead. The mechanism you proposed would work the same, just ignoring the additional fields the value (since the invoice has already been updated to its final state).

It could be a simple boolean option internally (possibly exposed via rpc as well) that asks for intermediate states as well

Reply:
I was just considering the trade-off between implementation effort and functionality. I agree that deterministic behaviour is better. What is an RSM btw?

You are right about the abstraction level. These are two independent things.

And yes, the diff in definition of `update_index` isn't much. It is a new bucket anyway.

Okay, so we go for replayable history.

Just wondering if we want to keep the dynamic fields in the invoice as well. The alternative is only store static fields in the invoice and use the invoice field `update_index` to retrieve the latest dynamic values.

In any case, I think if we store intermediate values, we should also stream them over rpc.

Reply:
I added a commit for the `InvoiceEvent` message.

Reply:
Catching up on this conversation, I think this is similar to what I described above, in that the API can be purely event based so the user doesn't need to filter out fields which each update, instead we give them a discrete event to handle.

Reply:
Filter out, as in ignore all static invoice fields because they are already known?